camel/memories/blocks/graphdb_block.py

```diff
@@ -0,0 +1,176 @@
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+import json
+import logging
+import os
+from typing import Any, Dict, List, Optional
+
+import nltk
+from networkx import Graph
+
+from camel.memories.base import MemoryBlock
+from camel.memories.records import ContextRecord, MemoryRecord
+from camel.storages.graph_storages import BaseGraphStorage, Neo4jGraph
+
+logger = logging.getLogger(__name__)
+
+
+class GraphDBBlock(MemoryBlock):
+    r"""An implementation of the MemoryBlock abstract base class for
+    maintaining and retrieving information using a graph database.
+
+    Args:
+        storage (Optional[BaseGraphStorage], optional): The storage mechanism
+            for the graph database. Defaults to a specific implementation
+            if not provided. (default: None)
+    """
+
+    def __init__(
+        self,
+        storage: Optional[BaseGraphStorage] = None,
+        url: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        database: Optional[str] = "neo4j",
+    ) -> None:
+        self.storage = storage or self.default_graph_storage(
+            url, username, password, database
+        )
+
+    def default_graph_storage(
+        self,
+        url: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        database: Optional[str] = "neo4j",
+    ) -> BaseGraphStorage:
+        # Try to get details from environment variables first
+        url = os.getenv('NEO4J_URI', url)
+        username = os.getenv('NEO4J_USERNAME', username)
+        password = os.getenv('NEO4J_PASSWORD', password)
+        database = os.getenv('NEO4J_DATABASE', database)
+
+        # Log an error if any of the necessary values are missing
+        if not url or not username or not password:
+            logger.error(
+                "Neo4j connection detailed are missing."
+                "Ensure environment variables or parameters are set for"
+                "NEO4J_URI, NEO4J_USERNAME, and NEO4J_PASSWORD."
+            )
+            raise ValueError("Missing Neo4j connection details.")
+
+        # Create and return an instance of Neo4jGraph
+        return Neo4jGraph(url, username, password, database or "neo4j")
+
+    def retrieve(
+        self, query: str, params: Optional[Dict[str, Any]] = None
+    ) -> List[ContextRecord]:
+        r"""Retrieves records from the graph database based on a query.
+
+        Args:
+            query (str): The query string to execute.
+            params (Optional[Dict[str, Any]]): Parameters for the query.
+
+        Returns:
+            List[ContextRecord]: A list of context records retrieved from the
+                graph database.
+        """
+        try:
+            results = self.storage.query(query, params)
+            scores = self._calculate_scores(query, results)
```

Comment:
here the query is some query language like Cypher, calculate the score between the query and the matched result from graph database seems doesn't make sense